observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemAdd.java

```diff
@@ -0,0 +1,99 @@
+import static org.jboss.as.server.deployment.Phase.POST_MODULE_METRICS;
+import static org.wildfly.extension.micrometer.MicrometerSubsystemDefinition.MICROMETER_COLLECTOR;
+import static org.wildfly.extension.micrometer.MicrometerSubsystemExtension.SUBSYSTEM_NAME;
+
+import java.util.List;
+import java.util.function.Function;
+
+import org.jboss.as.controller.AbstractBoottimeAddStepHandler;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.registry.ImmutableManagementResourceRegistration;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.as.server.AbstractDeploymentChainStep;
+import org.jboss.as.server.DeploymentProcessorTarget;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.ServiceController;
+import org.wildfly.extension.micrometer.deployment.MicrometerSubsystemDeploymentProcessor;
+import org.wildfly.extension.micrometer.metrics.MicrometerCollector;
+import org.wildfly.extension.micrometer.metrics.MicrometerCollectorService;
+
+public class MicrometerSubsystemAdd extends AbstractBoottimeAddStepHandler {
+    MicrometerSubsystemAdd() {
+        super(MicrometerSubsystemDefinition.ATTRIBUTES);
+    }
+
+    public static final MicrometerSubsystemAdd INSTANCE = new MicrometerSubsystemAdd();
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void performBoottime(OperationContext context, ModelNode operation, ModelNode model)
+            throws OperationFailedException {
+        super.performBoottime(context, operation, model);
+
+        // If the MP Metrics module is not installed, we need to install the WF Metrics DPU and initiate a metrics
+        // collection. If MP Metrics *is* installed, then we do not need to do either of those things, as that module
+        // handles that instead.
+        if (!context.getCapabilityServiceSupport().hasCapability("org.wildfly.extension.metrics.scan")) {
+            List<String> exposedSubsystems = MicrometerSubsystemDefinition.EXPOSED_SUBSYSTEMS.unwrap(context, model);
+            boolean exposeAnySubsystem = exposedSubsystems.remove("*");
+            boolean securityEnabled = MicrometerSubsystemDefinition.SECURITY_ENABLED.resolveModelAttribute(context, model)
+                    .asBoolean();
+
+            MicrometerRegistryService.install(context, securityEnabled);
+            MicrometerCollectorService.install(context);
+            MicrometerContextService.install(context, securityEnabled);
+
+            context.addStep(new AbstractDeploymentChainStep() {
+                @Override
+                public void execute(DeploymentProcessorTarget processorTarget) {
+                    processorTarget.addDeploymentProcessor(SUBSYSTEM_NAME, DEPENDENCIES, 0x1910,
+                            new MicrometerDependencyProcessor());
+                    processorTarget.addDeploymentProcessor(SUBSYSTEM_NAME, POST_MODULE, POST_MODULE_METRICS - 1, // ???
```

Comment:
I don't think we have had any discussion that drops the need for all phases to be added to the Phase class - if we end up with duplicate phase IDs I am not sure how well the behaviour is defined.

In the past it is not a problem to send a PR to WildFly Core to reserver the ID and then just hard code the ID value in WildFly and later update to the constant.

On this topic I would like at some point to look into how we can firstly clean up the DUP phases, and secondly automatically handle the dependency ordering.  Regarding the ordering if we could minimise / eliminate interaction within a single phase then the order they run within the phase would become less relevent.

Reply:
> I don't think we have had any discussion that drops the need for all phases to be added to the Phase class 

Right, I'm not ware of such a discussion either and I didn't wanted sound categoric on my statement. It was more a general way to do it because is what I've been seeing in the code. I agree with your comment about how to proceed with it.

Reply:
@jasondlee If you haven't already, please send a PR to core to reserve the necessary Phase id.

I'm ok with proceeding with this with a follow-up commit to move to that id once a Core release with it is merged.

Reply:
I've created this PR: https://github.com/wildfly/wildfly-core/pull/5222